[Docs] Update documentation for the new GitHub workflow

tru opened the pull request:
This describes the GitHub workflow in the documentation and marks some sections as deprecated. We should clean up these sections ASAP. I was just keen to get something on the documentation site as soon as possible.

tru committed:
[Doc] Update documentation for the new GitHub workflow

This describes the GitHub workflow in the documentation
and marks some sections as deprecated. We should clean up
these sections ASAP.

PiotrZSL commented on llvm/docs/Contributing.rst:
```diff
@@ -201,5 +182,6 @@ of LLVM's high-level design, as well as its internals:
 .. _git-clang-format: https://reviews.llvm.org/source/llvm-github/browse/main/clang/tools/clang-format/git-clang-format
-.. _LLVM's Phabricator: https://reviews.llvm.org/
+.. _LLVM's GitHub: https://github.com/llvm/llvm-project
+.. _LLVM's Phabricator (soon deprecated): https://reviews.llvm.org/
```
Just call it "deprecated", as basically it's already is.

tru replied:
Done.

jh7370 commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
+
+  gh pr create
+
+When promted select to create and use your own fork and follow
```
Typo: "promted" -> "prompted"

mstorsjo commented on llvm/docs/GettingInvolved.rst:
```diff
@@ -110,7 +105,7 @@ Discourse forums. There are also commit mailing lists for all commits to the LLV
 
 `LLVM Discourse`__
-  The forums for all things LLVM and related sub-projects. There are categories and subcategories for a wide variety of areas within LLVM. You can also view tags or search for a specific topic. 
+  The forums for all things LLVM and related sub-projects. There are categories and subcategories for a wide variety of areas within LLVM. You can also view tags or search for a specific topic.
```
There seems to be a lot of these fixups of trailing whitespace. It's probably fine, but mildly distracting when looking for the actual changes. (It would probably be fine to just push a commit that cleans up that whitespace without review.)

nickdesaulniers replied:
+1 trailing whitespace and line wraps

jh7370 commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+The button that should be used is called `Squash and merge` and after you can
+select the option `Delete branch` to delete the branch from your fork.
+
+You can also merge via the CLI by switch to your branch locally and run:
```
"switch" -> "switching"

jh7370 commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+
+Checking out another PR locally
+-------------------------------
+Sometimes you want to review another persons PR on your local machine to run
```
"persons" -> "person's"

bjope commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+-------------------
+When your PR has been accepted you can use the web interface to land your change.
+The button that should be used is called `Squash and merge` and after you can
+select the option `Delete branch` to delete the branch from your fork.
```
What happens to the review history when the branch is removed? Is it still possible to go back and follow up on all the review comments five years later if the branch is removed, or is all that stored somewhere else.

And if it is stored somewhere else, then I don't understand how rebases of the branch in my fork would "make the context of the old changes and comments harder to find and read". Or is that always the case anyhow when doing the final "squash and merge". In Phabricator I think comment line up nicely if for example looking at a specific diff from the patches history (i.e. a diff matching with the timeline when the comment was written).

felipepiovezan replied:
The review is always available on github.

> I don't understand how rebases of the branch in my fork would "make the context of the old changes and comments harder to find and read".

I should have said, the "last state of the review" is always available on github. When we force-push (which is pretty much required after a rebase), previous states are gone forever. Github only knows how to track your PR against *one* base; if we rebase and things change, some comments may be dangling

bjope replied:
Ok, thanks for explaining. But then the --delete part should be ok!

jh7370 commented on llvm/docs/MyFirstTypoFix.rst:
```diff
@@ -345,6 +345,12 @@ all the \*-commits mailing lists).
+.. warning::
+
+  Phabricator is deprecated and will be switched to read-only mode in October
+  2023, for new code contributions use :ref:`GitHub Pull Requests <github-reviews>`.
```
The "for new" bit should be a separate sentence. (i.e. "... 2023. For new...")

PiotrZSL reviewed (commented):
Overall ok, but raises even more questions:
- Will pushing new commits to branch with open pull request in my fork require some manual "work" on pull request, or new changes to my branch will become immediately visible on pull request.
- How about when committer != author on forked branch, who will be used as an author when squash + merge when someone else click "merge" (lets assume single commit pull request).
- How about pushing changes without merge, just cherry-pick commit from pull request.
- How about "labels" for push requests, who should manage them ?

jh7370 commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
+The LLVM project is using GitHub Pull Requests for Code Reviews. This document
+describes the typical workflow of creating a Pull Request and getting it reviewed
+and accepted. This is meant as an overview of the GitHub workflow, for compelete
+documentation refer to `GitHubs documentation <https://docs.github.com/pull-requests>`_.
```
GitHubs -> GitHub's

jh7370 reviewed (commented):
No objections here, other than some minor typos/grammar errors etc.

mstorsjo commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+Landing your change
+-------------------
+When your PR has been accepted you can use the web interface to land your change.
+The button that should be used is called `Squash and merge` and after you can
```
When using "Squash and merge", does it preserve the commit messages, or at least give a chance to edit the final commit message? This is quite vital. With the given procedures, I would expect the first commit in the branch to have the intended commit message, and later commits to be fixups. In such a case, we'd want to maintain the full commit message of the first commit, and skip the commit messages of the later fixup commits.

Dropping the commit message of the individual commits and summarizising it into a list of commits subjects like this, would be entirely unacceptable:

> Merge PR #1234
>
> Change the frobnicator
>
> Fixup
>
> Fixup

kbeyls replied:
Agreed, we probably need some guidance here on how to ensure a good commit message get's associated with the commit that ends up landing on main.

IIRC, github will offer an editable text box in the webui that contains the concatenation of all commit messages in the PR. Assuming the first commit contains a well-written commit message, and the follow-ons are simply fixups, it should be easy to simple remove the "fixup" commit messages.

kbeyls replied:
(As an aside: if it would take too long to research the details needed for writing proper documentation, I think it would be fine to write a placeholder TODO/FIXME in the documentation, e.g. stating that we need more documentation on how to ensure a good commit message, and land this PR, so that the documentation that is ready becomes available sooner rather than later).

tru replied:
> I have no clue what it does when you have multiple commits in a branch and opening a PR, I've not done this before and we should probably not encourage anyone doing that.

I think it takes the title and the body from the first commit in that branch to use as a title and commit message.

zero9178 commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+
+Rebasing Pull Requests and Force Pushes
+---------------------------------------
+In general, you should avoid rebasing a Pull Request and force pushing to the
+branch that's the root of the Pull Request during the review. This action will
+make the context of the old changes and comments harder to find and read.
```
Isn't this too restrictive? Not rebasing your branch during review is too restrictive IMO. You might have conflicts with `main` at some point, so you have to rebase at some point to land it and it makes more sense for a reviewer to review the post-rebase changes in my opinion. 
Additionally, in such case I am assuming CI will fail as well, as it can only run if there are no conflicts with `main` right? Having to accept a review that is not tested by CI seems odd to me.

Would it make more sense to say "do not invalidate the commit history by squashing and ammeding commits", but that you can rebase and then force-push as needed? Or maybe this should be in the "Updating Pull Requests" section.

tru replied:
Not putting much value in this (I think I agree more with you), but the closest thing to a consensus I could find was that people wanted to avoid rebases.

joker-eph replied:
I don't think this is a good advice here, unless you explicitly say: "you  should merge `origin/main` into your branch regularly".

Otherwise you may run into conflicts, or mid-air collision: any `ninja check` you run locally isn't testing against the latest `main` either.
It's not even clear what the pre-merge CI would test: would it be an outdated version of the code?

nhaehnle replied:
I think *avoiding* rebases is a good policy (though I'd encourage folks to try my [diff-modulo-base](https://git.sr.ht/~nhaehnle/diff-modulo-base) tool to help in reviews), but it should be *allowed*.

That most closely reflects how things work with Phabricator.

zero9178 commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
+----------------------
 When updating a pull request, you should push additional "fix up" commits to
 your branch instead of force pushing.  This makes it easier for GitHub to
 track the context of previous review comments.
```
(Cannot comment on line 70 in the new diff, so this is the next closest thing 🙃 )

The advice to "squash and merge before committing" is very confusing to me. Is this supposed to be for landing and is already superseded by your "Landing your change" section?

It reads to me like it contradicts the previous parahraph that said "you should just push additional commits to update your PR, don't force push" followed by "squash and merge before commiting" (which requires a force push). 
Should that paragraph just be deleted?

tru replied:
I think it should say `squash and merge before landing your PR`. I have updated the text now.

felipepiovezan commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
 
+Creating Pull Requests with GitHub CLI
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+With the CLI it's enough to create the branch locally and then run:
```
Should we be explicit that the original PR creation *must always* have a single commit?
If our advice is to squash and merge, there is an implication here that all PRs must start as a single commit.

This is the scenario I'm worried about: a well intentioned developer creates a lot of small patches (great!) but make a single PR, because they are used to other GitHub flows. The current advice will destroy all the nice work done to create the small commits.

mstorsjo replied:
Yes, this is also my concern. For this reason, I strongly prefer rebasing and squashing in fixes and forcepushing for each update of the branch. I know github has been rather bad at tracking old review comments across force pushes before, but I have a faint feeling it's working better these days. (And if we are going to squash at the end, or if we do allow the occasional rebase anyway when needed, we will lose context anyway.)

Personally, when submitting a PR, I would be very vocal about this, about how I want it merged (squashed, or kept as is, or if we insist on incremental fixes in separate commits, I would say that I want to do the final squashing once things are approved, and only then let someone merge it, with rebase+merge instead of squash+merge).

tru replied:
See my comment below, I think there is a valid discussion to have here, but not sure this is the right place to have it. We need to get the documentation in, and what's described here is the **closest** thing to the consensus I have found.

felipepiovezan replied:
Just to make sure we are on the same page, I am not proposing anything new or opposing the squash and merge flow; this is what the majority of people involved in the discussion agreed upon.
The rationale for that decision (AFAICT) was to mimic the phabricator workflow as closely as possible, and that workflow starts on the premise that *one* unit of contribution is reviewed through a Differential review. To emphasize this intent, I do believe that our new documentation should state that "PRs should be created from a single commit", it is a 1:1 match to what exists in phabricator today.

tru replied:
How about adding this:

```
Keep in mind that each pull request should generally only contain one commit.
This makes it easier for reviewers to understand the introduced changes and
provide feedback. It also helps maintain a clear and organized commit history
for the project. If you have multiple changes you want to introduce, it's
recommended to create separate pull requests for each change.
```

felipepiovezan replied:
Looks great! :)

joker-eph replied:
Note we also have a section that advise against amend+force-push for the PR workflow ("Updating Pull-request" below)

tstellar replied:
We currently only allow squash and merge for pull requests, I think this is a good policy because it makes it harder to make a mistake and accidentally push a series of fixup commits.  Even with this setting though, it's still possible to have a pull request with multiple commits from the start, you just have to push it manually once it's reviewed rather than going through the PR interface.

felipepiovezan replied:
> , you just have to push it manually once it's reviewed rather than going through the PR interface.

This is a great point and I am glad it is a possibility, but it seems like an advanced use for the following reasons:

* This is likely not what most GH users are used to doing, so I think the current wording is appropriate.
* GH will not close the review automatically and there won't be any "closed by commit X" messages like we get in phab. Manual intervention is required here.
*  This doesn't play well with the advice of not force-pushing (we usually want to address comments *on the commit* they were mentioned, so that the log is preserved properly after merge).
* CI won't be testing each commit to ensure they all compile / test.

felipepiovezan commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+
+  # When your PR is accepted, you can now rebase it and make sure
+  # you have all the latest changes.
+  git rebase -i origin/main
```
In the original [discourse post](https://discourse.llvm.org/t/update-on-github-pull-requests/71540/53), I believe you had said:

> This is easiest done with git rebase --interactive origin/main and mark the commits with f for fixup

I feel we should emphasize the fixup here and remove the `--squash` from the merge command.
Squashing through github will create that concatenation of commit messages.

tru replied:
I have updated the wording in the landing fixes section now. Does it feel better to you?

karka228 commented on llvm/docs/GitHub.rst:
```diff
@@ -46,11 +73,106 @@ The default commit message for a squashed pull request is the pull request
+
+  # Commit, use a good commit message
+  git commit file.cpp
+
```
Maybe add a comment about building and running some testcases also.

tru replied:
Yeah maybe - I was not adding it here because it's more covered by the Contributing document. But if nothing else we should link to it from here.

joker-eph replied:
It is there after 6b6312b01d1f now

kbeyls commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
+
+.. note::
+
+  When you let the GitHub CLI to create a fork of llvm-project to
```
I'm a non-native speaker, so I could be wrong, but I think "When you let the GitHub CLI create a fork" is easier to understand (i.e. drop the "to").

pogo59 replied:
This native speaker agrees.

kbeyls reviewed (commented):
I noticed that there are a few more files in the llvm/docs directory that mention phabricator, besides the ones that you touched in this patch. I think it's fine to leave updating those for a later/separate PR. It seems you've covered the most important ones here.

tru commented:
Reading the comments here, the most controversial thing is the rebase / squash and merge flow. To be clear, this is how the project is currently configured - if you go to a PR and try to merge it, it will only allow you to use the "Squash and merge" flow - so I also used that wording in the document.

Going back and reading some of the discussion, [here](https://discourse.llvm.org/t/enable-contributions-through-pull-request-for-llvm/53627) and [here](https://discourse.llvm.org/t/code-review-process-update/63964)  about the migration I think squash and merge was chosen for this task because it most closely mimics the workflow we are already using with Phab. 

I think this policy can be discussed to allow another flow, but I am not sure this documentation (in the last second) PR is the right forum.

My suggestion is that we keep the wording about "squash and merge" but add some wording about you can either:

* rebase interactively and fixup the commits and the commit message before finally landing the PR.
* use squash and merge, but please remember to edit your commit message when using the convenient button in the web UI.

Would that be a good first draft of this document and then we can have the rest of the discussion on discourse to alter this policy in the future?

pogo59 commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
-policies LLVM will be adopting once the project starts using them.
+The LLVM project is using GitHub Pull Requests for Code Reviews. This document
+describes the typical workflow of creating a Pull Request and getting it reviewed
+and accepted. This is meant as an overview of the GitHub workflow, for compelete
```
s/compelete/complete/

tru commented:
> Overall ok, but raises even more questions:
> 
> * Will pushing new commits to branch with open pull request in my fork require some manual "work" on pull request, or new changes to my branch will become immediately visible on pull request.

It will be automatically updated when you push to your branch. I will add that into the document.

> * How about when committer != author on forked branch, who will be used as an author when squash + merge when someone else click "merge" (lets assume single commit pull request).

GitHub will handle that, you as a committer just press the button in the web-ui and github will use the right authorship, even when squashing iirc.

> * How about pushing changes without merge, just cherry-pick commit from pull request.

Unsure what you mean? doing the merge manually from the command line? I think that's something you can do, but I rather we try to use the flow that most people are used to (pressing buttons in the web interface).

> * How about "labels" for push requests, who should manage them ?

Unclear, I have no thoughts on that at this point.

pogo59 commented on llvm/docs/GitHub.rst:
```diff
@@ -21,21 +24,45 @@ branches being used for "stacked" pull requests will be allowed.
+You can interact with GitHub in several ways: via git command line tools,
+the web browser, `GitHub Desktop <https://desktop.github.com/>`_, or the
+`GitHub CLI <https://cli.github.com>`_. This guide will cover the git command line
+tools and the GitHub CLI. The GitHub CLI (`gh`) will be most like the `arc` workflow and
```
I see examples with `gh` but not using just `git` itself.

tru replied:
I have left links to documentation for the normal `git` command where applicable. Is there some flow that's only documented with `gh` below that is possible with `git` that I have missed?

joker-eph replied:
This has been done now

pogo59 reviewed (commented):
Interestingly, I see various comments marked as resolved but I don't see the updates in the PR.

tru commented on llvm/docs/CodeReview.rst:
```diff
@@ -78,10 +78,8 @@ author, for example, might no longer be an active contributor to the project.
 ------------------------------------
 
-Pre-commit code reviews are conducted on our web-based code-review tool (see
-:doc:`Phabricator`). Post-commit reviews can be done on Phabricator, by email
```
Fixed!

tru commented on llvm/docs/Contributing.rst:
```diff
@@ -122,28 +122,9 @@ For more information on LLVM's code-review process, please see :doc:`CodeReview`
-  % git log origin/main...HEAD --oneline
-  # Push to Github.
-  % git push origin HEAD:main
+Once a patch is reviewed, you can select the "Squash and merge" button in the
```
We currently only have squash and merge enabled. See my comment below.

nhaehnle replied:
I assume this means manual pushes are still acceptable, especially in the context of multi-commit sequences. That is probably worth stating explicitly somewhere.

tru commented:
> Interestingly, I see various comments marked as resolved but I don't see the updates in the PR.

I am marking them as resolved as I am fixing them locally. I haven't pushed the update yet - lot of comments here!

tru committed:
Code review updates.

Clarified how to use squash and merge and other smaller fixes.

jh7370 commented:
> > Interestingly, I see various comments marked as resolved but I don't see the updates in the PR.
> 
> I am marking them as resolved as I am fixing them locally. I haven't pushed the update yet - lot of comments here!

FWIW, I think it should be up to the reviewer to mark a thing as resolved, because otherwise GitHub "conveniently" hides the review comment, and the reviewer finds it hard to double-check to see if the comment has actually been resolved (they have to go through and expand every "resolved" comment to see their previous comment, one of my many gripes about how bad GitHub is for a reviewer experience versus Phabricator). However, that's probably a discussion that needs to happen on Discourse, to determine a community-wide policy.

tru committed:
Another rephrase

unterumarmung commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,173 @@ branches being used for "stacked" pull requests will be allowed.
+  # fork and upstream to llvm/llvm-project
+  git push origin my_change
+
+  # When your PR is accepted, you can now rebase it and make sure
+  # you have all the latest changes.
+  git rebase -i origin/main
```
Is it actually a necessary step? GitHub allows merging PRs without requiring the PR to be on ToT as long as there are no merge conflicts with ToT. I'm not certain if this is also the case with GitHub CLI, but it certainly is with GitHub Web. I assume they should function in the same way.

Additionally, the original Phabricator documentation included a reminder to run tests before merging. I believe that if we suggest rebasing, we should also recommend running tests before merging.

To clarify, I'm not insisting; I'm just offering ideas on how to improve the documentation. Thank you for your time!

joker-eph replied:
It does not change anything about merge conflict (if the PR can be merged by Github, there isn't any), not rebasing affect the kind of test you're running locally!
(that said we're having pre-merge CI? It may do the rebase/merge when it runs?)

unterumarmung commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,173 @@ branches being used for "stacked" pull requests will be allowed.
+  # Commit, use a good commit message
+  git commit file.cpp
+
+  # Create the PR, select to use your own fork when prompted.
+  gh pr create
```
If a user does not have a fork, then will there be an error? So we do not pollute the main repository with the branches

tru replied:
No it will create the fork for you. You will get a clear dialog to chose what to do.

unterumarmung replied:
Nice, thanks!

jh7370 commented:
> > FWIW, I think it should be up to the reviewer to mark a thing as resolved, because otherwise GitHub "conveniently" hides the review comment, and the reviewer finds it hard to double-check to see if the comment has actually been resolved (they have to go through and expand every "resolved" comment to see their previous comment, one of my many gripes about how bad GitHub is for a reviewer experience versus Phabricator). However, that's probably a discussion that needs to happen on Discourse, to determine a community-wide policy.
> 
> That's a valid discussion to have I think. It's not been the workflow I have ever used on GitHub and there is no good way to enforce it, so I am not sure I agree - but let's have that discussion on Discourse.

I've gone and created it as https://discourse.llvm.org/t/rfc-github-pr-resolve-conversation-button/73178.

bjope commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,173 @@ branches being used for "stacked" pull requests will be allowed.
+  git rebase -i origin/main
+
+  # Now merge it
+  gh pr merge --squash --delete
```
So this is where I normally (in the phabricator days) would do

1) git pull
2) see what new stuff I got
3) if something might conflict and cause build failures etc, rebuild
4) if something might conflict and cause test failures etc, rerun tests
5) if step 3 and 4 took some time, go back to 1
6)  git push origin HEAD:main

So my goal is to push something that has little risk of breaking things in some obvious way that would have been detected by building and rerunning lit tests.
 
 I guess I'll still be able to do exactly the above thing. But then I guess the PR isn't updated automativally (which happened with Phabricator), or is it? Instead I need to remeber to push via my fork somehow (e.g. using "gh pr merge"). Still, I really want to keep the time here short so that I've verified my patch together with top-of-tree.  So I hope that "gh pr merge" is rather quick (or that it would fail if I'm not rebased to main making the merge a fast forward).

joker-eph replied:
Does `gh pr merge`  asks interactively about the resulting commit message? I think when I click "squash and merge" in the UI it pops-up and asks.

Our settings should be (hopefully) to use the PR description, but that would mean that amending a commit locally and setting a commit message would be ignored by `gh pr merge` ...

tru committed:
Added reminders about running tests, and links to other relevant documentation

tru committed:
Add paragraph about keeping one pr == one commit

tru commented:
I think I have addressed most of the comments in the discussion above, so if everyone can look again, try to point out any glaring problems. I think it's better that we get this merged when we turn on PR's with some issues than to have the old documentation still on our site.

sdesmalen-arm commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,188 @@ branches being used for "stacked" pull requests will be allowed.
+
+  # When your PR is accepted, you can now rebase it and make sure
+  # you have all the latest changes.
+  git rebase -i origin/main
```
Is this missing a step to update the fork with the latest commits in the llvm-project repo?

tru replied:
Yep will fix!

joker-eph replied:
This is in now.

joker-eph commented on llvm/docs/DeveloperPolicy.rst:
```diff
@@ -46,7 +46,7 @@ quality.
+Developers should stay informed by reading the `LLVM Discourse forums`_.
 If you are doing anything more than just casual work on LLVM, it is suggested that you also
 subscribe to the "commits" mailing list for the subproject you're interested in,
 such as `llvm-commits
```
We should mention the new teams to subscribe to and link to the CODEOWNERS file I think.

joker-eph commented on llvm/docs/GettingInvolved.rst:
```diff
@@ -130,9 +125,9 @@ The :doc:`CodeOfConduct` applies to all these forums and mailing lists.
 
 `LLVM Announcements`__
-  If you just want project wide announcements such as releases, developers meetings, or blog posts, then you should check out the Announcement category on LLVM Discourse. 
+  If you just want project wide announcements such as releases, developers meetings, or blog posts, then you should check out the Announcement category on LLVM Discourse.
```
Nit: line wrap (if you're already fixing formatting here...)

joker-eph commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+built-in tool. See the section about landing your fix below.
+
+When pushing to your branch, make sure you push to the correct fork. Check your
+remotes with:
```
Can you incorporate in this section the feedback and changes from https://reviews.llvm.org/D150594 ?

joker-eph commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+your branch instead of force pushing. This makes it easier for GitHub to
+track the context of previous review comments. Consider using the
+`built-in support for fixups <https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---fixupamendrewordltcommitgt>`_
+in git.
```
How is this making a difference for the PR / GitHub? Any link to GitHub doc?

joker-eph replied:
Ping on this @tra : I'm happy to help addressing post-review comment, but here I don't know why we have this?

tra replied:
> Ping on this @tra : I'm happy to help addressing post-review comment, but here I don't know why we have this?

I think you meant to ping @tru

joker-eph replied:
Yes, sorry @tra ! Thanks for redirecting.

joker-eph reviewed (commented):
Thanks!

tru commented:
Heads up - I have no time to work on this during the weekend. So either someone can merge this now and we open up more PRs for further discussion and changes or we merge this next week. I have no strong feeling either way.

joker-eph commented:
Thanks @tru, it's fine to follow-up on my comments and other after the week-end: having Phabricator still documented as "the way to go" is worse than having an imperfect GitHub documentation in the meantime.

nickdesaulniers commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+  # Create your changes
+  $EDITOR file.cpp
+
+  # Don't forget clang-format
+  git clang-format
```
So I think it's better to do:

`git clang-format HEAD~` after you commit.

Or `git clang-format HEAD~N` after `N` commits.

joker-eph replied:
This is what it is saying now after 6b6312b01d1f

nickdesaulniers commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
-For pull requests, please push a branch to your
-`fork <https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks>`_
+----------------------
+Keep in mind that each pull request should generally only contain one commit.
```
This seems to contradict advice later in the page about updating commits to contains multiple commits rather than rebasing.

Maybe this can be reworded slightly to clarify that the introduction of a new pull request should only contain one commit, but that more commits may be added to address code review feedback?

joker-eph replied:
Done in 9818bf2619b1

nickdesaulniers commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+Afterwards you can select the option `Delete branch` to delete the branch
+from your fork.
+
+You can also merge via the CLI by switching to your branch locally and run:
+
+::
+
+  gh pr merge --squash --delete-branch
```
maybe worth mentioning that you should probably rerun the relevant tests in between rebasing and merging.

`ninja check-<whatever>`.

joker-eph replied:
This is added now

nickdesaulniers commented on llvm/docs/Contributing.rst:
```diff
@@ -122,28 +121,9 @@ For more information on LLVM's code-review process, please see :doc:`CodeReview`
-  # Check that the list of commits about to be pushed is correct.
-  % git log origin/main...HEAD --oneline
-  # Push to Github.
-  % git push origin HEAD:main
```
L165 below talks about a pre-push hook; but the diff removed above removes the recommended workflow about `git push` and the new Github doc recommends using `gh pr merge`.

Two thoughts:
1. Does the pre-push hook still run if using the `gh` CLI?  If not, we might need to clean that up.
2. Can we link to the Github page's section on merging "Landing your change" from this page's "For developers to commit changes from Git" section?

hubert-reinterpretcast commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+`Interactive rebase <https://git-scm.com/docs/git-rebase#_interactive_mode>`_
+with fixup's. This is the recommended method since you can control the final
+commit message and inspect that the final commit looks as you expect. When
+your local state is correct, remember to force-push to your branch and press
```
This method does not insert the PR number into the commit subject line (unlike the other method). I thought having the PR number is considered best practice.

joker-eph replied:
This is still pushing to a fork, the sentence ends with "press the merge button".

What's misleading here is that it makes you believe the commit message you have locally matters: it won't, the message is taken from the PR description (and includes the PR number)

hubert-reinterpretcast replied:
Ah. Maybe this should say force-push to your branch and then push to main (because having control of the commit message locally can be valuable). I swear I had GitHub discard edits I made to the commit message in the web UI when "merging" the PR.

tru replied:
I still need to fix up this documentation. Hope to have time / energy soon :)

felipepiovezan replied:
>  PR number into the commit subject line 

Is this something we agreed to do / discussed? IMO this pollutes the `log --one-line` output, but a link at the end of the commit message is very valuable (like what we get with `arc diff`).

tru replied:
It's just what GitHub does - automatically inserts it at the end of the line.

hubert-reinterpretcast commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
 
-If you do this, you must squash and merge before committing and
+If you do this, you must squash and merge before landing the PR and
 you must use the pull request title and description as the commit message.
```
What about the PR number? This can be read as actively disallowing the inclusion of the PR number.

hubert-reinterpretcast commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+branch that's the root of the Pull Request during the review. This action will
+make the context of the old changes and comments harder to find and read.
+
+Sometimes, a rebase might be needed to update your branch with a fix for a test
```
Doesn't _merging_ the updated state of `main` into your branch work too (and not require a force push)?

joker-eph replied:
This is something I asked in https://reviews.llvm.org/D150594 and we didn't resolve about which way we should advise this!

bjope commented on llvm/docs/GitHub.rst:
```diff
@@ -21,36 +24,194 @@ branches being used for "stacked" pull requests will be allowed.
+
+  # When your PR is accepted, you can now rebase it and make sure
+  # you have all the latest changes.
+  git rebase -i origin/main
```
Since "gh" is changing the remotes ("origin" is made pointing at the fork) I had assumed that this step would involving "git rebase -i upstream/main" to rebase to the latest from "llvm"  and not my fork. To allow running tests and ensure that my patch is working with top-of-tree (that might impact what I want to say in the commit message etc).

(Is it perhaps possible that someone else could have edited my pull request so that I want to fetch things from my fork as well?)

joker-eph replied:
I think 6b6312b01d1f should make it clear

nickdesaulniers commented:
Follow up on how to rebase when `git pr merge` fails: https://github.com/llvm/llvm-project/pull/66223 RFC
